### PR TITLE
Improve handling timeouts when resolving shell env

### PR DESCRIPTION
**Pull request by gzdunek**
Closes: https://github.com/gravitational/webapps.e/issues/298

The main problem is that we don't have an access to the `NotificationService` (which lives in `AppContext`) from the preload script. 
~~A workaround (that I made) is to create an event emitter that `AppContext` will subscribe to. What do you think about it?~~

~~However, this will solve the current problem, but not the problem in general. What if the main process needs to send a notification? What if the shared process also has to? I have a feeling that each part of the application will use a different technique for this (IPC, gRPC, EventEmitter). What comes to my mind is to move the notification sending logic to the shared process, so both the main process and preload script could use it.
But for now it sounds like designing a really overcomplicated solution to just send a message from preload to the frontend.~~

### CODE REVIEW

**Comment by gzdunek** [comment]
In the new solution `createPtyProcess` returns status from the `buildPtyOptions`, so the renderer process can display a notification when there is a timeout.
This will make `preload` and `renderer` less coupled and won't disrupt the flow of code execution.
Thanks @ravicious for the idea.

**Comment by gzdunek** [comment]
In the new solution `createPtyProcess` returns status from the `buildPtyOptions`, so the renderer process can display a notification when there is a timeout.
This will make `preload` and `renderer` less coupled and won't disrupt the flow of code execution.
Thanks @ravicious for the idea.

File: packages/teleterm/src/services/pty/ptyHost/resolveShellEnv.ts [hunk 1]
```diff
@@ -104,11 +106,7 @@ async function resolveUnixShellEnv(
     abortSignal.onabort = () => {
       child.kill();
       logger.warn('Reading shell env timed out');
-      reject(
-        new Error(
-          'Unable to resolve shell environment. Please review shell configuration.'
-        )
-      );
+      reject(reject(new ResolveShellEnvTimeoutError()));
```
Comment (ravicious) [inline packages/teleterm/src/services/pty/ptyHost/resolveShellEnv.ts:109]:
```suggestion
      reject(new ResolveShellEnvTimeoutError());
```

File: packages/teleterm/src/services/pty/ptyHost/resolveShellEnv.ts [hunk 2]
```diff
@@ -46,10 +46,12 @@ import { spawn } from 'child_process';
 import { memoize } from 'lodash';
 
 const logger = new Logger('resolveShellEnv()');
-const resolveShellMaxTime = 8000; // 8s
```
[the remark this answers is not in the document]
Comment (gzdunek) [reply packages/teleterm/src/services/pty/ptyHost/resolveShellEnv.ts:49]: Hmm, I thought 8s looks a bit weird (why such an exact value?), people are more used to round values like 5s or 10s. But if is okay, I can bring it back :)

File: packages/teleterm/src/services/pty/ptyHost/buildPtyOptions.ts [hunk 3]
```diff
@@ -1,20 +1,54 @@
...
+    }))
+    .catch(error => {
+      if (error instanceof ResolveShellEnvTimeoutError) {
+        return {
+          shellEnv: undefined,
+          creationStatus: PtyProcessCreationStatus.ResolveShellEnvTimeout,
+        };
+      }
+      throw error;
+    });
```
[the remark this answers is not in the document]
Comment (gzdunek) [reply packages/teleterm/src/services/pty/ptyHost/buildPtyOptions.ts]: What would you say for rewriting the whole function to a regular promise?
I like it more in this case, `try-catch` requires us to use mutable variables.

File: packages/teleterm/src/ui/DocumentTerminal/useDocumentTerminal.ts [hunk 4]
```diff
@@ -110,6 +108,28 @@ async function initState(
...
+): Promise<IPtyProcess> {
+  try {
+    const { process, creationStatus } =
+      await ctx.terminalsService.createPtyProcess(cmd);
+
+    if (creationStatus === PtyProcessCreationStatus.ResolveShellEnvTimeout) {
+      ctx.notificationsService.notifyWarning({
+        title: 'Unable to resolve shell environment',
+        description:
+          'In order to source the environment variables the new shell session is opened, but the script took more than 10 seconds to finish. ' +
```
[the remark this answers is not in the document]
Comment (gzdunek) [reply packages/teleterm/src/ui/DocumentTerminal/useDocumentTerminal.ts]: Done

File: packages/teleterm/src/ui/DocumentTerminal/useDocumentTerminal.ts [hunk 5]
```diff
@@ -110,6 +108,28 @@ async function initState(
...
+          'This most likely means that your shell start up took longer to execute or it waits for an input during startup.',
```
[the remark this answers is not in the document]
Comment (gzdunek) [reply packages/teleterm/src/ui/DocumentTerminal/useDocumentTerminal.ts]: Done

[the remark this answers is not in the document]
Comment (gzdunek) [reply packages/teleterm/src/ui/DocumentTerminal/useDocumentTerminal.ts]: That will make the notification really long :(

What about the following:
title: Environment variables may be incomplete,
description: In order to source the environment variables, a new temporary shell session is opened and then immediately closed, but it didn't close within 10 seconds. 
This most likely means that your shell startup took longer to execute or that your shell waits for an input during startup. 
Please check your startup script.

File: packages/teleterm/src/services/pty/ptyHost/buildPtyOptions.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ravicious) [reply packages/teleterm/src/services/pty/ptyHost/buildPtyOptions.ts]: Between those two choices, I say let's leave it as it is. 😄

Personally I dislike how easy it is to be imprecise with error handling when using promises. I can see how having to use `let` with try-catch made you use promises here, though to be honest I don't think that there's anything wrong with redefining local vars like that.

Edit: Ah, you changed it already, I think I've opened some old tab with the PR or something. Let's use the promise-only version then.

File: packages/teleterm/src/ui/DocumentTerminal/useDocumentTerminal.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ravicious) [reply packages/teleterm/src/ui/DocumentTerminal/useDocumentTerminal.ts]: I like the idea with putting the effect of the failure in the title. Though what does it really mean for env vars to be incomplete? I don't think it's a common way to express this.

Maybe "Could not source environment variables for shell session"?

In "Please check your startup script.", I'd change "script" to "files".

**Review by ravicious — APPROVED** [review]
I like your ideas around the notification title and I understand now the need to change 8s to 10s.

File: packages/teleterm/src/ui/DocumentTerminal/useDocumentTerminal.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (gzdunek) [reply packages/teleterm/src/ui/DocumentTerminal/useDocumentTerminal.ts]: Done

**Pull request merged**